generic variadic arrays creates invalid c code

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->
**V doctor:**
```
OS: linux, "Fedora release 33 (Thirty Three)"
Processor: 4 cpus, 64bit, little endian, Intel(R) Celeron(R) CPU  N3160  @ 1.60GHz
CC version: cc (GCC) 10.3.1 20210422 (Red Hat 10.3.1-1)

getwd: /home/notecomm/pv/data_structures
vmodules: /home/notecomm/.vmodules
vroot: /home/notecomm/v
vexe: /home/notecomm/v/v
vexe mtime: 2021-07-08 09:52:50
is vroot writable: true
is vmodules writable: true
V full version: V 0.2.2 1e6230a.9ba6c23

Git version: git version 2.31.1
Git vroot status: weekly.2021.27-17-g9ba6c23d
.git/config present: true
thirdparty/tcc status: thirdparty-linux-amd64 7543de81
```

**What did you do?**
`v -g -o vdbg cmd/v && vdbg t.v`
```v
struct Node<T> {
mut:
  data T
  next &Node<T> = 0
}

struct SinglyLinkedList<T> {
  mut: first_node &Node<T> = 0
}

fn init_singlylinkedlist<T>(nodes ...Node<T>) SinglyLinkedList<T> {
  mut current_node := &nodes[0]

  for i in 0..nodes.len - 1 {
    current_node = &nodes[i]
    current_node.next = &nodes[i + 1]
  }

  return SinglyLinkedList{&nodes[0]}
}

fn main() {
  sll := init_singlylinkedlist<int>(Node<int>{data: 1}, Node<int>{data: 2}, Node<int>{data: 798})
  println(sll.first_node.next)
}
```

**What did you expect to see?**

```
&Node<int>{data: 2, next: &<circular>}
```

**What did you see instead?**
```
==================
/tmp/v/t.10344519151083520712.tmp.c:10220: error: ')' expected (got "<")
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

If you were not working with C interop, this is a compiler bug, please report the bug using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang

```
```c
// /tmp/v/t.10344519151083520712.tmp.c:10220
VV_LOCAL_SYMBOL void main__main(void) {
	main__SinglyLinkedList_T_int sll = main__init_singlylinkedlist_T_int(new_array_from_c_array(3, 3, sizeof(main__Node<>), _MOV((main__Node<>[3]){(main__Node_T_int){.data = 1,.next = 0,}, (main__Node_T_int){.data = 2,.next = 0,}, (main__Node_T_int){.data = 798,.next = 0,}})));
	println(str_intp(1, _MOV((StrIntpData[]){{_SLIT("&"), 0xfe10 ,{.d_s=main__Node_T_int_str(*sll.first_node->next)}}})));
}
```

--- a/vlib/v/gen/c/fn.v
+++ b/vlib/v/gen/c/fn.v
@@ -1173,12 +1173,10 @@ fn (mut g Gen) call_args(node ast.CallExpr) {
 		mut arr_info := arr_sym.info as ast.Array
 		if varg_type.has_flag(.generic) {
 			if fn_def := g.table.find_fn(node.name) {
-				varg_type_name := g.table.type_to_str(varg_type)
-				for i, fn_gen_name in fn_def.generic_names {
-					if fn_gen_name == varg_type_name {
-						arr_info.elem_type = node.concrete_types[i]
-						break
-					}
+				if utyp := g.table.resolve_generic_to_concrete(arr_info.elem_type, fn_def.generic_names,
+					node.concrete_types)
+				{
+					arr_info.elem_type = utyp
 				}
 			} else {
 				g.error('unable to find function $node.name', node.pos)

--- a/vlib/v/tests/generics_with_variadic_generic_args_test.v
+++ b/vlib/v/tests/generics_with_variadic_generic_args_test.v
@@ -1,0 +1,31 @@
+struct Node<T> {
+mut:
+	data T
+	next &Node<T> = 0
+}
+
+struct SinglyLinkedList<T> {
+mut:
+	first_node &Node<T> = 0
+}
+
+fn init_singlylinkedlist<T>(nodes ...Node<T>) SinglyLinkedList<T> {
+	mut current_node := &nodes[0]
+
+	for i in 0 .. nodes.len - 1 {
+		current_node = &nodes[i]
+		current_node.next = &nodes[i + 1]
+	}
+
+	return SinglyLinkedList{&nodes[0]}
+}
+
+fn test_generic_with_variadic_generic_args() {
+	sll := init_singlylinkedlist<int>(Node<int>{ data: 1 }, Node<int>{
+		data: 2
+	}, Node<int>{
+		data: 798
+	})
+	println(sll.first_node.next)
+	assert sll.first_node.next.data == 2
+}

